posthog/async_migrations/migrations/0004_replicated_schema.py

```diff
@@ -0,0 +1,305 @@
+import re
+from dataclasses import dataclass
+from functools import cached_property
+from typing import List, Optional
+
+import structlog
+from constance import config
+from django.conf import settings
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.sql.table_engines import MergeTreeEngine
+from posthog.async_migrations.definition import AsyncMigrationDefinition, AsyncMigrationOperation
+
+logger = structlog.get_logger(__name__)
+
+"""
+Migration summary:
+
+Schema change to migrate tables to support replication and more than
+one shard.
+
+This allows for higher scalability as more hosts can be added under ClickHouse.
+
+The migration strategy:
+
+    1. We have a list of tables that might need replacing below.
+    2. For each one, we replace the current engine with the appropriate Replicated by:
+        a. creating a new table with the right engine and identical schema
+        b. temporarily stopping ingestion to the table
+        c. using `ALTER TABLE ATTACH/DROP PARTITIONS` to move data to the new table.
+        d. rename tables
+        e. re-enabling ingestion
+
+We use ATTACH/DROP tables to do the table migration instead of a normal INSERT. This method allows
+moving data without increasing disk usage between identical schemas.
+
+`events` and `session_recording_events` require extra steps as they're also sharded:
+
+    1. The new table should be named `sharded_TABLENAME`
+    2. Create `TABLENAME` and `writable_TABLENAME` tables which are responsible for distributed reads and writes
+    3. Update materialized views to write to `writable_TABLENAME`
+
+Constraints:
+
+    1. This migration relies on there being exactly one node when it's run.
```

Comment:
Is this enforced in anyway? What happens if we run concurrently?

Reply:
Oh, wait maybe I misunderstood, this is one clickhouse node?

Reply:
It is in #validate_number_of_nodes_in_cluster

Note this is about clickhouse cluster size not parallel celery processes

Reply:
Improved the comment